assets/scripts/streets/StreetName.jsx

```diff
@@ -58,33 +110,40 @@ class StreetName extends React.PureComponent {
 
   renderHoverPrompt = () => {
     if (this.props.isStreetReadOnly || !this.props.isHoverable) return null
-    if (typeof this.props.onClick === 'function' && this.state.isHovered) {
+    if (typeof this.props.handleStreetNameChange === 'function' && this.state.isHovered) {
       return (
         <div className="street-name-hover-prompt">
           {t('street.rename', 'Click to rename')}
         </div>
       )
     }
-
     return null
   }
 
   render () {
     let classString = 'street-name-text ' + (!needsUnicodeFont(this.props.name) ? '' : 'fallback-unicode-font')
-
-    return (
-      <div
+    if (!this.state.modify) {
+      return (<div
         className="street-name"
         ref={this.props.childRef}
         onMouseEnter={this.onMouseEnter}
         onMouseLeave={this.onMouseLeave}
-        onClick={this.props.onClick}
-        id={this.props.id}
-      >
+        onClick={() => this.setState({modify: true})}
+        id={this.props.id}>
         {this.renderHoverPrompt()}
         <div className={classString}>{StreetName.normalizeStreetName(this.props.name)}</div>
-      </div>
-    )
+      </div>)
+    } else if (this.state.modify && this.props.editable) {
+      return (<div
+        className="street-name"
+        ref={this.props.childRef}
+        id={this.props.id}>
+        <input className={classString} type="text"
```

Comment:
The `<input>` element changes the UI too much, and it's a jarring experience. Have you considered using the [`contenteditable` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/contenteditable) on the `<div>` element itself? Alternatively, if `<input>` is preferred, the element should be styled so that it matches the original street sign appearance.

Either way, it may be necessary to override the width of the editable text container during the editable state.

Secondly, it is preferable to show a very slight change in the UI appearance to indicate editing. I will think about this more, but for now, try a border color change or a background color change.

Reply:
How about adding two buttons at the top like the cancel button in the 'my streets' section? One for dismissing changes and the other for saving them

Reply:
I don't think we should be introducing that level of complexity for editing in place. If we need more UI, we could consider scrapping the edit-in-place UI altogether, and use a modal instead. (It would basically be re-implenting the browser's native prompt functionality.)

Reply:
So input is a no-no, should we move to contentEditable? You did say you were looking at other alternatives.

Reply:
I wanted to do a little bit of research and try to find some references that might help think about the editing state UI.

1. Here is a very basic example of a `contentEditable` element for a street plaque: https://codepen.io/louh/pen/doVqeK
2. Relevant tips about the ["edit in place" pattern in _Designing Interfaces_](http://designinginterfaces.com/firstedition/index.php?page=Edit-in-Place):

> Make sure the text field appears in precisely the same apparent location as the original non-editable string. If it seems to jump when editing starts, it may irritate users. (This situation doesn't seem like a big deal, but it can be.) Also, retain the same typeface as the original text, if you can. In short, make it as WYSIWYG as possible.

> Usually a border appears around the text when the editor is invoked, indicating to the user that editing has begun. That may not even be necessary, though. Other cues may suffice: a text-entry cursor should appear (often blinking), and if a common user task is to replace the original string entirely, then the whole original string should be automatically selected when the editor appears.

3. Example: using a change in border color to indicate editing state:

![](https://cdn-images-1.medium.com/max/1600/1*zKU-XEeYddX24yMtf6xCAA.gif)

4. Example: using a change in background color to indicate editing state:

![](https://cdn-images-1.medium.com/max/1600/1*J0T8qCSKrGo27R_5_UDZVw.gif)

5. Example: using a change in page overlay to indicate focusing on the editable area:

![](https://cdn-images-1.medium.com/max/1600/1*EowNqfbkeEVXPWrqwR6r5w.gif)

6. Example that uses animation while expanding width to show a change in state:

![](https://cdn-images-1.medium.com/max/1600/1*OmBxhrucDHn-bbaLU6qROQ.gif)

In the prior attempt, the input transitioned instantaneously, which led to it feeling too jarring when it changed from the view to edit state. You mentioned that you liked the longer state as a distinction. I think if we wanted to give size a try, the animation of that transition is very important. In addition, we should make sure that the line-height of the text is preserved after the transition.

7. In most of the examples I found, having a "confirmation" button of some kind seems very normal. Although I didn't like the clutter, it does seem that the visual trigger might be necessary.

Reply:
All these points are very interesting -- the second one is particularly helpful. I think the best approach would be to prepare a mockup of these to compare? If we agree I can start  working on these, though I'll be able to present them only after my mid sems (they end on 13th)